tests/components/airvisual_pro/test_config_flow.py

```diff
@@ -0,0 +1,61 @@
+"""Test the AirVisual Pro config flow."""
+from unittest.mock import patch
+
+from pyairvisual.node import NodeProError
+import pytest
+
+from homeassistant import data_entry_flow
+from homeassistant.components.airvisual_pro.const import DOMAIN
+from homeassistant.config_entries import SOURCE_USER
+from homeassistant.const import CONF_IP_ADDRESS, CONF_PASSWORD
+
+
+async def test_duplicate_error(hass, config, config_entry):
+    """Test that errors are shown when duplicates are added."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=config
+    )
+    assert result["type"] == data_entry_flow.FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
+
+
+@pytest.mark.parametrize(
+    "exc,errors",
+    [
+        (NodeProError, {"base": "cannot_connect"}),
+        (Exception, {"base": "unknown"}),
+    ],
+)
+async def test_errors(hass, config, exc, errors):
+    """Test that an exceptions show an error."""
+    with patch(
+        "homeassistant.components.airvisual_pro.config_flow.NodeSamba.async_connect",
+        side_effect=exc,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER}, data=config
+        )
+        assert result["type"] == data_entry_flow.FlowResultType.FORM
+        assert result["errors"] == errors
```

Comment:
We should finish the flow all the way to either a CREATE or ABORT result

Reply:
Unclear on this. Doesn't the flow stop at `FORM` when errors occur?

Reply:
We should finish the flow all the way to a final result to make sure it can recover from invalid entry.  Feed it good values after the bad values to get to either create or abort.